supernode/services/userdataprocess/task.go

```diff
@@ -0,0 +1,528 @@
+	log.WithContext(ctx).Debugf("signAndSendSNDataSigned begin to sign SuperNodeRequest")
+	signature, err := task.pastelClient.Sign(ctx, []byte(sndata.UserdataHash+sndata.UserdataResultHash), task.config.PastelID, task.config.PassPhrase, "ed448")
+	if err != nil {
+		return errors.Errorf("failed to sign sndata %w", err)
+	}
+	if !isPrimary {
+		sndata.HashSignature = hex.EncodeToString(signature)
+		sndata.NodeID = task.config.PastelID
+		log.WithContext(ctx).Debug("send signed sndata to primary node")
+		if _, err := task.ConnectedTo.ProcessUserdata.SendUserdataToPrimary(ctx, sndata); err != nil {
+			return errors.Errorf("failed to send signature to primary node %s at address %s %w", task.ConnectedTo.ID, task.ConnectedTo.Address, err)
+		}
+	}
+	return nil
+}
+
+// AddPeerSNDataSigned gather all the Userdata with signature from other Supernodes
+func (task *Task) AddPeerSNDataSigned(ctx context.Context, snrequest userdata.SuperNodeRequest) error {
+	log.WithContext(ctx).Debugf("addPeerSNDataSigned begin to aggregate SuperNodeRequest")
+
+	task.peersSNDataSignedMtx.Lock()
+	defer task.peersSNDataSignedMtx.Unlock()
+
+	var actionErr error
+	<-task.NewAction(func(ctx context.Context) error {
+		log.WithContext(ctx).Debugf("receive supernode userdata result signed from node %s", snrequest.NodeID)
+		if node := task.accepted.ByID(snrequest.NodeID); node == nil {
+			actionErr = errors.Errorf("node %s not in accepted list", snrequest.NodeID)
+			return nil
+		}
+
+		task.peersSNDataSigned[snrequest.NodeID] = snrequest
+		if len(task.peersSNDataSigned) == len(task.accepted) {
+			log.WithContext(ctx).Debug("all signature received")
+			go func() {
+				close(task.allPeersSNDatasReceived)
+			}()
+		}
+		return nil
+	})
+	return actionErr
+}
+
+func (task *Task) verifyPeersUserdata(ctx context.Context) (userdata.ProcessResult, error) {
+	log.WithContext(ctx).Debugf("all supernode data signed received so start validation")
+	successCount := 0
+	dataMatchingCount := 0
+	for _, sndata := range task.peersSNDataSigned {
+		// Verify the data
+		signature, err := hex.DecodeString(sndata.HashSignature)
+		if err != nil {
+			log.WithContext(ctx).Debugf("failed to decode signature %s of node %s", sndata.HashSignature, sndata.NodeID)
+			continue
+		}
+
+		ok, err := task.pastelClient.Verify(ctx, []byte(sndata.UserdataHash+sndata.UserdataResultHash), string(signature), sndata.NodeID, "ed448")
+		if err != nil {
+			log.WithContext(ctx).Debugf("failed to verify signature %s of node %s", sndata.HashSignature, sndata.NodeID)
+			continue
+		}
+
+		if !ok {
+			log.WithContext(ctx).Debugf("signature of node %s mistmatch", sndata.NodeID)
+			continue
+		}
+
+		successCount++
+		if task.ownSNData.UserdataHash == sndata.UserdataHash &&
+			task.ownSNData.UserdataResultHash == sndata.UserdataResultHash {
+			dataMatchingCount++
+		}
+	}
+
+	if successCount < task.config.MinimalNodeConfirmSuccess-1 {
```

Comment:
Actually if we don't need to wait for all the confirmation then confirmation can by aggregated one by one and can end early when reach `task.config.MinimalNodeConfirmSuccess-1`.

Reply:
In the future there is requirement to set reputation score for the one that cannot confirm success, so we get all confirmation for now, so in the future we know which one not confirm.